File: from-books/hash-table-exercises/language/python/exercise-3.py
# Write a function that acceps a string that contains all the letters of the alphabet
# except one and returns the missing letter. For example, the string "the quick brown box
# jumps over a lazy dog" contains all the letters of the alphabet except the letter, "f".
# The function should have a time complexity of O(n).
def generateAlphabet() -> dict[str, int]:
    alphabet: dict[str, int] = {}

    for i in range(26):
        letter: str = chr(97 + i)
        alphabet[letter] = i + 1

    return alphabet


def missingLetter(sentence: str) -> str:
    alphabetHashMap: dict[str, int] = generateAlphabet()
    sentenceMap: dict[str, int] = {}

    for letter in sentence:
        if not sentenceMap.get(letter):
            sentenceMap[letter] = 1

    missing: str = ""
    for letter in alphabetHashMap:
        if not sentenceMap.get(letter):
            missing = letter
            break

    return missing


print(missingLetter("the quick brown box jumps over a lazy dog"))

File: from-books/hash-table-exercises/language/go/exercise-3.go
package main

import "fmt"

func main() {
	checkString := "the quick brown box jumps over a lazy dog"

	result := findMissingLetter(checkString)

	fmt.Println(result)
}

// Write a function that acceps a string that contains all the letters of the alphabet
// except one and returns the missing letter. For example, the string "the quick brown box
// jumps over a lazy dog" contains all the letters of the alphabet except the letter, "f".
// The function should have a time complexity of O(n).
func generateAlphabet() map[string]int {
	alphabet := make(map[string]int, 0)

	for i := 0; i < 26; i++ {
		char := string(rune(97 + i))
		alphabet[char] = i + 1
	}

	return alphabet
}

func findMissingLetter(sentence string) string {
	alphabetHashMap := generateAlphabet()
	sentenceMap := make(map[string]int, 0)

	for _, letter := range sentence {
		if _, exists := sentenceMap[string(letter)]; !exists {
			sentenceMap[string(letter)] = 1
		}
	}

	missingLetter := ""
	for letter := range alphabetHashMap {
		if _, exists := sentenceMap[letter]; !exists {
			missingLetter = letter
			break
		}
	}

	return missingLetter
}
